include/multipass/exceptions/not_implemented_on_this_platform_exception.h

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (C) 2020 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef MULTIPASS_NOT_IMPLEMENTED_ON_THIS_PLATFORM_EXCEPTION_H
+#define MULTIPASS_NOT_IMPLEMENTED_ON_THIS_PLATFORM_EXCEPTION_H
+
+#include <stdexcept>
+
+#include <multipass/format.h>
+
+namespace multipass
+{
+class NotImplementedOnThisPlatformException : public std::runtime_error
+{
+public:
+    NotImplementedOnThisPlatformException(const std::string& feature)
+        : runtime_error(fmt::format("The {} feature is not implemented on the {} platform.", feature,
+#ifdef MULTIPASS_PLATFORM_LINUX
+                                    "Linux"
+#else
+#ifdef MULTIPASS_PLATFORM_APPLE
+                                    "MacOS"
+#else
+                                    "Windows"
+#endif
+#endif
+                                    ))
```

Comment:
I don't think we should do this, the exception `NotImplementedOnThisBackendException` is just fine for networking stuff in virtualbox-linux (before your other PR). Even if it is implemented in virtualbox-mac/win. And we should definitely avoid proliferating ifdefs around. Platform dependent code should be contained in the platform source as much as possible. 

No point adding another exception type anyway. If all you wanted was a different message you could have thrown `std::runtime_error`. We're not catching it or treating it specially, so the only advantage of a separate exception type is to quickly identify the condition. But then having both `NotImplementedOnThis{Backend,Platform}Exception` only muddles that.

Perhaps think of `NotImplementedOnThisBackendException` as short(-ish) for `NotImplementedOnThisBackendPlatformCombinationException`. It's not like we're defining the term "backend" formally anyway. And conceptually it makes sense to view it as including any platform specificities that it depends on. That's what the vbox code is about: the behavior depends on platform specificities. But we could imagine a backend that didn't. Let "xyz" be a *backend* that does not need to call the platform code in question. When you do, on Linux, `mp set local.driver=xyz && mp lsn`, it does not throw. But when you then do `mp set local.driver=virtualbox && mp lsn`, it now throws, even though the platform didn't change. It is only reasonable to say that `lsn` is not implemented *on that backend*. I would even say that, if platform_linux threw NotOnThisPlatform, the calling backend would need to convert it to NotOnThisBackend, which is the right thing to tell its caller. Of course there's no point in any of that.

Reply:
Thinking about it from an orthogonal point of view, that of the user, we can consider the combination backend/platform only as the backend. Thus, I agree with your point, thank you for the clear explanation.
I'll make it `NotImplementedOnThisBackendException`, that also simplifies things.

Reply:
OK great, thank you. The full include guard that you had was a good idea though (I hadn't noticed it before). In case we add another `NotImplemented...Exception` in the future.